packages/contracts-bedrock/scripts/Artifacts.s.sol

```diff
@@ -21,11 +22,134 @@ struct Deployment {
     address payable addr;
 }
 
+interface VmReadCallers is VmSafe {
+    function readCallers() external view returns (VmSafe.CallerMode callerMode, address msgSender, address txOrigin);
+    function startBroadcast(address signer) external view;
+    function stopBroadcast() external view;
+    function broadcast(address signer) external view;
+}
+
 /// @title Artifacts
 /// @notice Useful for accessing deployment artifacts from within scripts.
 ///         When a contract is deployed, call the `save` function to write its name and
 ///         contract address to disk. Inspired by `forge-deploy`.
+/// @dev This uses the DeploymentRegistry contract, to load/store deployments,
+///      without broadcasting the interactions with the DeploymentRegistry.
+///      The DeploymentRegistry can be overridden, to load/store deployments in a sub-context, without file IO.
 abstract contract Artifacts {
+
+    /// @notice Foundry cheatcode VM.
+    Vm private constant vm = Vm(address(uint160(uint256(keccak256("hevm cheat code")))));
+    VmReadCallers private constant vmSafe = VmReadCallers(address(uint160(uint256(keccak256("hevm cheat code")))));
+
+    Artifacts public constant deploymentRegistry =
+        Artifacts(address(uint160(uint256(keccak256(abi.encode("optimism.deploymentregistry"))))));
+
+    function setUp() public virtual {
+        if (!vm.envBool("DISABLE_DEPLOYMENT_REGISTRY")) {
+            vm.etch(address(deploymentRegistry), vm.getDeployedCode("Artifacts.s.sol:DeploymentRegistry"));
+            vm.label(address(deploymentRegistry), "Artifacts");
+            vm.allowCheatcodes(address(deploymentRegistry));
+            deploymentRegistry.setUp();
+        }
+    }
+
+    /// @notice Modifier that wraps a function to temporarily not broadcast.
+    function _noBroadcastPre() internal view returns (VmSafe.CallerMode callerMode_, address msgSender_){
+        VmSafe.CallerMode callerMode;
+        address msgSender;
+        address txOrigin;
+        (callerMode, msgSender, txOrigin) = vmSafe.readCallers();
+        // If we were broadcasting, stop it.
+        if (callerMode == VmSafe.CallerMode.RecurrentBroadcast || callerMode == VmSafe.CallerMode.Broadcast) {
+            vmSafe.stopBroadcast();
+        }
+        return (callerMode, msgSender);
+    }
+
+    function _noBroadcastPost(VmSafe.CallerMode _callerMode, address _msgSender) internal view {
+        // Now recover the broadcasting state, if any
+        if(_callerMode == VmSafe.CallerMode.RecurrentBroadcast) {
+            vmSafe.startBroadcast(_msgSender);
+        }
+        if(_callerMode == VmSafe.CallerMode.Broadcast) {
+            vmSafe.broadcast(_msgSender);
+        }
+    }
```

Comment:
You may want to just make this a modifier:
```solidity
/// @notice Modifier that wraps a function to temporarily not broadcast.
modifier noBroadcast() {
    VmSafe.CallerMode callerMode;
    address msgSender;
    address txOrigin;
    (callerMode, msgSender, txOrigin) = vmSafe.readCallers();
    
    // If we were broadcasting, stop it.
    if (callerMode == VmSafe.CallerMode.RecurrentBroadcast || callerMode == VmSafe.CallerMode.Broadcast) {
        vmSafe.stopBroadcast();
    }
    
    // Execute the function.
    _ ;

    // Now recover the broadcasting state, if any
    if(_callerMode == VmSafe.CallerMode.RecurrentBroadcast) {
        vmSafe.startBroadcast(_msgSender);
    }
    if(_callerMode == VmSafe.CallerMode.Broadcast) {
        vmSafe.broadcast(_msgSender);
    }
}
```

Example usage:
```solidity
// Before
function save(string memory _name, address _deployed) public {
    (VmSafe.CallerMode callerMode, address msgSender) = _noBroadcastPre();
    deploymentRegistry.save(_name, _deployed);
    _noBroadcastPost(callerMode, msgSender);
}

// After
function save(string memory _name, address _deployed) public noBroadcast {
    deploymentRegistry.save(_name, _deployed);
}
```

Reply:
I tried that, but it didn't work for `view` functions. However, since those are non-broadcasting, it might actually work as non-view modifier.